### PR TITLE
Utf8String LastIndexOf() 

### PR DESCRIPTION
**Pull request by A-And**
Related to #1712 and #1718 for Feature request #1510 
Implemented for substrings, codepoints and code units.
Apologies for the larger PR.

### CODE REVIEW

File: tests/System.Text.Utf8String.Tests/RandomTests.cs [hunk 1]
```diff
@@ -631,6 +631,37 @@ public void IndexOfTests(int expected, string s, string substring)
         }
 
         [Theory]
+        [InlineData(0, "", "")]
+        [InlineData(0, "abc", "")]
```
Comment (KrzysztofCwalina) [inline tests/System.Text.Utf8String.Tests/RandomTests.cs]: I tried it on .NET ( ```c# Console.WriteLine("abc".LastIndexOf(""));``` ) and it prints 2. I don't think it makes sense, but maybe we should be compatible? @stephentoub, @terrajobst, @joshfree, any opinions on this?

File: tests/System.Text.Utf8String.Tests/RandomTests.cs [hunk 2]
```diff
@@ -631,6 +631,37 @@ public void IndexOfTests(int expected, string s, string substring)
...
+        [InlineData(38, "aaaaaaabaaaaaaaabaaaaaaaabaaaaaaaabaaaaaaaaaaab", "aaaaaaaab")]
+        [InlineData(16, "abbbbbbbabbbbbbbabbbbbbbbbbbbbbbbbbbbbbbbbbbb", "abbbbbbbb")]
+        [InlineData(-1, "aaabaaa", "aabaaaa")]
+        [InlineData(-1, "aaabaaa", "aadaaaa")]
+        [InlineData(3, "ababababa", "bababa")]
+        [InlineData(-1, "abababab", "bbababa")]
+        [InlineData(-1, "abababab", "bababaa")]
+        [InlineData(7, "aaabaacaaac", "aaac")]
+        [InlineData(9, "baaaabaaabaaaa", "baaaa")]
+        [InlineData(0, "\uABC0\uABC1\uABC2", "\uABC0\uABC1")]
```
Comment (KrzysztofCwalina) [inline tests/System.Text.Utf8String.Tests/RandomTests.cs]: Could you add a case that uses codeponts from outside of the BMP, but one that does not return 0? Thanks!

File: tests/System.Text.Utf8String.Tests/RandomTests.cs (hunk 1 above)
Comment (danmosemsft) [reply tests/System.Text.Utf8String.Tests/RandomTests.cs]: Unless there's a good reason to not be compatible (eg significant perf impact) why would we not be intentionally compatible...

Comment (KrzysztofCwalina) [reply tests/System.Text.Utf8String.Tests/RandomTests.cs]: Because I am not sure how returning 2 makes any sense here. :-)

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test Innerloop Ubuntu16.04 Release Build and Test

**Comment by KrzysztofCwalina** [comment]
Thanks!

**Pull request merged**